### ISSUE TITLE
Meta-issue for new syntax additions and/or features

### ISSUE THREAD

**Issue by supposedly** [eltrhn/rueltabel#10]
- **Delayed auxiliary transitions:** Rare to find a use case, but allow something along the lines of a `+N` "operation" in an auxiliary to signify that it should take place only after N generations (the compiler would then step through the table to locate the appropriate cellstates that would result from allowing cells in that neighborhood to evolve untouched for N ticks).  
For example, in Brainfuck, the E part of the line `WEST_POINTER, W L_BRACK, NW..SE any, S (L_BRACK, prgm), SW any; [S: (BLANK_2, pointerOps)] -> W:L_BRACK  E[S: (NORTH_R_BRACK, _, ...)]` caused issues because a westbound pointer traveling over an L-bracket but not stopping at it would cause a NORTH_R_BRACK to be born. One solution is to remove the E and phrase it as another line, `any, SW L_BRACK, W BLANK_2, NW..S any; NORTH_R_BRACK` to indicate that the transition should take place only after the transition from WEST_POINTER -> BLANK_2 has occurred... but it could be expressed more concisely as a modification to the original line to the tune of `WEST_POINTER, W L_BRACK, NW..SE any, S (L_BRACK, prgm), SW any; [S: (BLANK_2, pointerOps)] -> W:L_BRACK  E+1[S: (NORTH_R_BRACK, _, ...)]`.
- **Bringing custom symmetries into Nutshell itself:** Currently this requires a Python extension class, but it could totally be specified in the table itself. Two big parts of it are Calcyman's suggestion from a while ago to the Golly Gang for indicating symmetries by specifying what napkins are to be equivalent, and BlinkerSpawn's `symmetry:`  notation for specifying which neighbor pairs the ~ operator permutes on.

**Comment by supposedly** [comment]
> _**Generalized references**: From HactarCE, extend the concept of bindings and mappings to make them usable anywhere in a transition and in particular within an operation as in (2, 3, 4)-[W] (meaning any variable in {2, 3, 4} excluding the state currently to the west). This coincides partially with #6 in that it too concerns a much-needed generalization of references. Both should be somewhat easier to implement once #9 is satisfied._

This has been resolved with #12.

**Comment by supposedly** [comment]
For **custom neighborhoods** and **bringing custom symmetries into Nutshell itself**, I'm thinking of a sort of poor-man's-YAML syntax within a new `@DEFINE` segment:

```
@DEFINE
neighborhood: some_nhood
some_nhood neighborhood: NW, NE, S

symmetries: BogusPermute
BogusPermute some_nhood:
- NW, NE, S
== S, NE, NW
- permute: NW, S
BogusPermute vonNeumann:
- N, E, S, W
== W, S, E, N
== S, W, N, E
- permute: N, S
- permute: N, E

symmetries: Whatever
Whatever Moore:
- N, NE, E, SE, S, SW, W, NW
== N, NW, W, SW, S, SE, E, NE

@TABLE
states: 12
symmetries: BogusPermute
neighborhood: some_nhood
```

or, more generally,

```
key: value
value subvalue: single value
value subvalue:
- list
of values
where each line is only counted as a new list item
if it starts with a hyphen
- like so
- else it's appended
to the last list item which did

key: value
value subvalue: etc
```
There's a bunch to figure out, though.
- Should the `permute:` within a list element as in the symmetry definitions count as a new key/value pair or just a list item that's to be parsed externally?
- Should newlines be ignored/converted to spaces in list items?
- Should the weird repetition of keys be ditched (instead just using indentation like in YAML?) -- maybe not

**Comment by supposedly** [comment]
> _**Auto-numbering of `@NUTSHELL` constants:** Maintaining numbers for cellstates is a pain and mostly an unnecessary burden to be forced upon the writer, considering that after they've defined constants they're probably not going to look at the numbers at all._

This has been resolved (and documented) with 99debef.

**Comment by supposedly** [comment]
> _**Statelist rotation:** This will be done with `>>` and `<<`, mimicking some languages' bitshift operators._

> _**Custom neighborhoods:** Could be neat to allow `neighborhood: [clockwise-ordered list of compass directions]` which would then (after symmetry normalization) be expanded into Moore or perhaps the smallest encompassing neighborhood. This requires symmetry classes' attributes to be reworked slightly, as currently their info is based on neighborhood length rather than specific positions of neighborhood cells (because none of Golly's neighborhoods overlap in length)._

These have respectively been resolved with 45e9725 and 4d37658 -- although the latter solution contains a glaring inefficiency, in that it goes through the for-loop recalculation of the appropriate Golly neighborhood for *every* transition when it only should need to do so once, which has yet to be fixed. 
 
(This was addressed, along with some other issues, in 4ab5ff3.)

> _**Inline rulestrings:** [...] This may have to wait for custom neighborhoods to be implemented, or more specifically for the "reworking" entailed thereby to be implemented -- because precautions will have to be taken here to prevent a Hensel configuration not valid in the chosen neighborhood from being used._

4d37658 also addresses this, and, more generally, d5acf99 brought inline-rulestring syntax into the grammar + standalone parser.

**Comment by supposedly** [comment]
> _**Inline rulestrings:** Syntax for defining a Hensel rulestring & fore-/background states instead of a normal transition napkin, which will come with an implicit `symmetries: rotate4reflect` for relevant neighborhoods. It will resemble `initial <hensel-rulestring / (foreground, states) / (background, states)> resultant`, where each of the fg/bg statelists can also be a single state on its own. Compass-direction references in the resultant state here will operate according to canonical Hensel orientation.  
> (There is also potential for introducing other hensel-notated rule systems, such as Deficient (even if it's not ever going to be used much); this could be accounted for by allowing the name thereof after the rulestring, e.g. `<hensel-rulestring algo-name / ...>`.)_

This is implemented as of f3fb7e6 (documented in c15511b).

**Comment by supposedly** [comment]
> _**References in more places:** Currently, references are allowed neither in the initial (input) term of a transition nor the background/foreground terms of an inline-rulestring transition._

Addressed in 71d2d69.

**Comment by supposedly** [comment]
Closing this as it's more-or-less identical to the "**Planned**" section of CHANGELOG.md (which had not been added at the time of this issue's creation).
